### ISSUE TITLE
PostgreSQL should emit connection strings

### ISSUE THREAD

**Issue by nevir** [hashicorp/vault#317]
It'd be valuable to have the PostgreSQL backend (and other database backends) emit full connection strings; maybe via a template.

Currently we've gotta wire together that information from multiple locations (e.g. retrieve the address & db name in consul, and user/pass from vault), even though vault should have most/all of the info required

maybe a `vault read MOUNT/connection/ROLE`?

**Comment by jefferai** [comment]
Adding to the 0.6 milestone as we have some work around postgres planned for that. I can't promise that this will actually be added, but I want to make sure that we don't forget about it so that it is in the discussion.

**Comment by vishalnayak** [comment]
If this issue is only about reading the configured parameters, it will be fixed by #1515

**Comment by vishalnayak** [comment]
Fixed by #1515

**Comment by jefferai** [comment]
I think this is a different ask -- the connection strings being asked for here will have information about the _generated_ user, not the higher level connection information valid for the static admin user.

**Comment by schuylr** [comment]
+1 but for MySQL backend, please. I noticed the issue labels don't cover this backend even though you closed another issue for MySQL concerning this as a duplicate.

**Comment by jefferai** [comment]
Issue labels don't really matter in this context :-)

**Comment by schuylr** [comment]
Making sure. Thanks :)

**Comment by stklcode** [comment]
What about adding an optional `additional_data` field to the `/<mount>/roles/<name>` endpoint expecting a key-value map.
When credentials are requested through `/<mount>/creds/<name>`, the `data` section is extended by those values. (not overwriting username/password)

This way you can configure not only URLs different from the internal connection, but also additional information like DB names or specific drivers and hence completely provision an application with a single request.

**Comment by serverhorror** [comment]
Could this be solved by a kind of templated output?

* Oracle (I know not in the backends but just to show): `SERVER(DESCRIPTION=(ADDRESS=(PROTOCOL=TCP)(HOST={{fqdn}})
    (PORT={{port}}))(CONNECT_DATA=(SID={{dbname}})));uid={{username}};pwd={{password}}`
* MySQL: `Server={{fqdn}};Database={{dbname}};Uid={{username}};Pwd={{password}};
SslMode={{sslmode}};`
* PostreSQL
  * URI Style: `postgresql://{{username}}@{{fqdn}}/{{dbname}}?sslmode={{sslmode}}`
  * Connection String: `host={{fqdn}} port={{port}} dbname={{dbname}} connect_timeout={{timeout}}`

I'm just putting different template words in here to show that it would be pretty generic. Vault could replace those with the values that are stored in the actual secret. I think that would apply to everything that can use connection strings.

In combination with some kind of output format that would be very powerful, then again one could generate that with a pipe thru `vault read -format=json .. | jq '.data....'` but that seems cumbersome.

**Comment by stklcode** [comment]
All those template approaches assume that the backend connection URI is the same as for external use. I think that's what @schuylr mentioned before.

Imagine the Vault server(s) connect through a local IP subnet or even Unix sockets a potential client does not have access to. (access limitations, performance, DNS issues ... whatever)

As the template solution requires a significant amoung of rework on current connection config and assuming one role is created for one type of service, I'd definitely plead for a static solution, that should be fine for most use cases.

**Comment by serverhorror** [comment]
Agreed the static is way simpler. I guess what I was having in my had was, in the case of PostgreSQL, something like this:

```
$ vault mount postgresql
$ vault write postgresql/config/connection connection_url="postgresql://root:vaulttest@vaulttest.ciuvljjni7uo.us-west-1.rds.amazonaws.com:5432/postgres" # only used to generate users, not for conn-strings
$ # this is where you can attach all the key/values to the secret like one can with the generic backend
$ vault write postgresql/config/lease \
    lease=1h \
    lease_max=24h \
    fqdn=a-fqdn \
    dname=a-dbname \
    sslmode=an-sslmode \
    port=a-port
$ # this is where you tell vault what template to use when an connection string is requested
$ vault write postgresql/config/template \
    template_string="postgresql://{{username}}:{{password}}@{{fqdn}}/{{dbname}}?sslmode={{sslmode}}"
$ vault write postgresql/roles/readonly \
    sql="CREATE ROLE \"{{name}}\" WITH LOGIN PASSWORD '{{password}}' VALID UNTIL '{{expiration}}';
    GRANT SELECT ON ALL TABLES IN SCHEMA public TO \"{{name}}\";"
$ vault read -format templated postgresql/creds/readonly
postgresql://root-1430162075-7887:34205e88-0de1-68b7-6267-72d8e32c5d3d@a-fqdn/a-dbname?sslmode=an-sslmode
```

I'm not good with naming so `-format templated` and such might be a bad choice but I hope the idea is somehow clearer now.

**Comment by adamdecaf** [comment]
What's the status of this? I'm indifferent on the format (kv pairs, jdbc url, etc) but having the hostname, and dbname would help a lot.

I'll help with the code or take a shot myself if there's a clear direction. 

This issue looks related too: https://github.com/hashicorp/vault/issues/3127

**Comment by jefferai** [comment]
@adamdecaf There is no status currently.

**Comment by adamdecaf** [comment]
Should I go ahead and propose something then? There are a few ideas in this thread already.

**Comment by adamdecaf** [comment]
I don't think vault should return JDBC/DSN, but perhaps the hostname, port, and database name?

We could check that `hostname` is non-loopback and not a local socket.

**Comment by serverhorror** [comment]
> We could check that hostname is non-loopback and not a local socket
Please don't!
How would you know the local layout of a random company using vault?
It might be intentional

**Comment by jefferai** [comment]
@Andrey9kin Not currently, there's very little demand - a handful of requests over the years. If someone wanted to work on a PR we could help with making sure the design is right.

**Comment by Andrey9kin** [comment]
@jefferai how hard could it be? ;) doesn't feel like something big. I guess extend plugins/helper/database/credsutil/sql.go to return one more value and then roll from there, right?

**Comment by Andrey9kin** [comment]
or perhaps it should be more specific since databases have a different format for connection strings...

**Comment by adamdecaf** [comment]
Yea, we never decided on a response format.

**Comment by therealgambo** [comment]
I think returning the `connection_url` of the database would be a decent start, not sure if handling the construction of the DSN should be done by vault or leave it to the application making the request.
